api/ocs/__init__.py

```diff
@@ -372,78 +478,248 @@ def parse_nonsidereal_target(data: dict) -> NonsiderealTarget:
+            return next(filter(lambda f: f in input, filter_dict), None)
+
+        if instrument == 'Visitor Instrument':
+            instrument = data[OcsProgramProvider._InstrumentKeys.NAME].split(' ')[0]
+            if instrument in ["'Alopeke", "Zorro"]:
+                fpu = None
+            else:
+                fpu = instrument
+        else:
+            if instrument in instruments_keys:
+                if instruments_keys[instrument] in data:
+                    fpu = data[instruments_keys[instrument]]
+                else:
+                    fpu = None
+                    # TODO: Might need to raise an exception here. Check code with science.
+            else:
+                raise ValueError(f'Instrument {instrument} not supported')
+
+        if OcsProgramProvider._AtomKeys.DISPERSER in data:
+            disperser = data[OcsProgramProvider._AtomKeys.DISPERSER]
+        elif instrument in ['IGRINS', 'MAROON-X']:
+            disperser = instrument
+        else:
+            disperser = None
+        
+        if instrument == 'GNIRS':
+            if data[OcsProgramProvider._InstrumentKeys.ACQ_MIRROR] == 'in' and data[OcsProgramProvider._InstrumentKeys.DECKER] == 'acquisition':
+                disperser = 'mirror'
+            else:
+                disperser = disperser.replace('grating', '') + data[OcsProgramProvider._InstrumentKeys.CROSS_DISPERSED]
+        elif instrument == 'Flamingos2' and fpu == 'FPU_NONE':
+            if data['instrument:decker'] == 'IMAGING':
+                disperser = data['instrument:decker']
+
+        if OcsProgramProvider._AtomKeys.FILTER in data:
+            filter = data[OcsProgramProvider._AtomKeys.FILTER]
+        elif instrument == 'GPI':
+            filter = find_filter(fpu, OcsProgramProvider.GPI_FILTER_WAVELENGTHS)
+        else:
+            if instrument == 'GNIRS':
+                filter = None
+            else:
+                filter = 'Unknown'
+        if instrument == 'NIFS' and 'Same as Disperser' in filter:
+            filter = find_filter(disperser[0], OcsProgramProvider.NIFS_FILTER_WAVELENGTHS)
+        wavelength = OcsProgramProvider.GPI_FILTER_WAVELENGTHS[filter] if instrument == 'GPI' else float(data[OcsProgramProvider._AtomKeys.WAVELENGTH])
+
+        return (fpu, disperser, filter, wavelength)
+    
     @staticmethod
     def parse_atoms(sequence: List[dict], qa_states: List[QAState]) -> List[Atom]:
         """
         Atom handling logic.
-
-        TODO: Update this with Bryan's newer code.
         """
-        n_steps = len(sequence)
-        n_abba = 0
+
+        def guide_state(step: dict) -> bool:
+            return any('guideWith' in key and guide == 'guide' for key, guide in step.items())
+        
+        def select_qastate(states: List[QAState]) -> QAState:
+            # Precedence order for observation classes.
+            qastate_order = [QAState.NONE, QAState.UNDEFINED, QAState.FAIL, QAState.USABLE, QAState.PASS]
```

Comment:
We don't need this (and furthermore, again, this list is being created every time the function is called): the `QAState` `enum` is actually an `IntEnum` that has this exact ordering on it, so we can just compare directly.